pkg/controller/core.oam.dev/v1alpha2/applicationconfiguration/applicationconfiguration.go

```diff
@@ -576,36 +574,80 @@ func (e *GenerationUnchanged) Error() string {
 		"Please ignore this error in other logic.")
 }
 
-func applyOnceOnly() apply.ApplyOption {
-	return func(ctx context.Context, current, desired runtime.Object) error {
-		if current == nil {
+// applyOnceOnly is an ApplyOption that controls the applying mechanism for workload and trait.
+// More detail refers to ApplyOnceOnlyMode
+func applyOnceOnly(ac *v1alpha2.ApplicationConfiguration, mode core.ApplyOnceOnlyMode) apply.ApplyOption {
+	return func(_ context.Context, existing, desired runtime.Object) error {
+		if mode == core.ApplyOnceOnlyOff {
 			return nil
 		}
-		// ApplyOption only works for update/patch operation and will be ignored
-		// if the object doesn't exist before.
-		c, _ := current.(metav1.Object)
+
 		d, _ := desired.(metav1.Object)
-		if c == nil || d == nil {
-			return errors.Errorf("invalid object being applied: %q ",
+		if d == nil {
+			return errors.Errorf("cannot access metadata of object being applied: %q",
 				desired.GetObjectKind().GroupVersionKind())
 		}
-		cLabels, dLabels := c.GetLabels(), d.GetLabels()
-		if dLabels[oam.LabelOAMResourceType] == oam.ResourceTypeWorkload ||
-			dLabels[oam.LabelOAMResourceType] == oam.ResourceTypeTrait {
-			// check whether spec changes occur on the workload or trait,
-			// according to annotations and lables
-			if c.GetAnnotations()[oam.AnnotationAppGeneration] !=
-				d.GetAnnotations()[oam.AnnotationAppGeneration] {
+		dLabels := d.GetLabels()
+		if dLabels[oam.LabelOAMResourceType] != oam.ResourceTypeWorkload &&
+			dLabels[oam.LabelOAMResourceType] != oam.ResourceTypeTrait {
+			// if target resource is not a workload nor trait
```

Comment:
You should give some example here, in which case will a resource applied that is not workload and trait

Reply:
I will remove this logic because currently, we use this function only for workloads and traits

Reply:
Maybe scope will use it？

Reply:
Scope still uses raw client Update but not 3-way-diff Apply.

Reply:
yes, maybe we will use it in the future, so I think we should keep the code but add an annotation to tell why( even the reason is some kinds of possiblilty)

Reply:
Got it.